src/numerics/include/metaphysicl/dualnumber_decl.h

```diff
@@ -470,6 +505,61 @@ template <typename T, typename D>
 inline
 D gradient(const DualNumber<T, D>& a);
 
+template <typename T, std::size_t N>
+struct DualNumberSurrogate;
+
+// Generic helper functions called when forwarding dual number method calls to the value and
+// derivatives that the dual number contains
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumber<R, NumberArray<N, R>> return_dn(const R & (TBase::*fn)(PtrArgs...) const,
+                                           const DualNumber<T, NumberArray<N, T>> & calling_dn,
+                                           ParamArgs &&... args);
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumber<R, NumberArray<N, R>> return_dn(R (TBase::*fn)(PtrArgs...) const,
+                                           const DualNumber<T, NumberArray<N, T>> & calling_dn,
+                                           ParamArgs &&... args);
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumberSurrogate<R, N> & return_dns(R & (TBase::*fn)(PtrArgs...),
+                                       DualNumber<T, NumberArray<N, T>> & calling_dn,
+                                       ParamArgs &&... args);
+
+template <typename T, typename TBase, size_t N, class... PtrArgs, class... ParamArgs>
+void const_void_helper(void (TBase::*fn)(PtrArgs...) const,
+                       const DualNumber<T, NumberArray<N, T>> & calling_dn,
+                       ParamArgs &&... args);
+
+template <typename T, typename TBase, size_t N, class... PtrArgs, class... ParamArgs>
+void void_helper(void (TBase::*fn)(PtrArgs...),
+                 DualNumber<T, NumberArray<N, T>> & calling_dn,
+                 ParamArgs &&... args);
+
+// Macros for declaring and defining dual number methods that will forward their arguments to the
+// underlying value and derivatives
+
+#define ConstReturnDecl(methodName)                                                                \
```

Comment:
Likewise with the other macro: change these to e.g. metaphysicl_const_return_decl to avoid name conflicts and match the existing MetaPhysicL style?

The design here may be slightly sacrificing readability for the sake of cleverness and brevity, but I don't suggest changing that at all; it [matches the existing MetaPhysicL style perfectly](https://www.youtube.com/watch?v=KUXb7do9C-w&t=16).